text/0000-module-unification.md

```diff
@@ -0,0 +1,837 @@
+creating your component module in
+`src/ui/routes/posts/post/-components/post-viewer.js`.
+
+#### Non-resolved Files
+
+The rules above apply to modules that are resolved, namely `*.js` and `*.hbs`
+files. Other files that are used for documenting code, such as `*.md` and
+`*.html` files, can be freely co-located in any directories.
+
+Conventions will still be used for non-resolved files that have significance
+within an Ember project, including:
+
+* `src/ui/styles` - A project's stylesheets.
+* `src/ui/index.html` - A project's html container.
+
+### Packages
+
+In-repo addons (including engines) will be placed in a new top-level `packages`
+directory (a sibling of `src`). We can begin to use the term "packages" instead
+of the rather clumsy "in-repo addons". This differentiation will emphasize that
+packages are internal and addons are external to a project. Packages should be
+seen as a lightweight way to add new namespacing within a project without the
+overhead of a full addon.
+
+The `packages` directory will provide a separate space away from other library
+modules that might be kept in `lib`, the current directory used for in-repo
+addons. Introducing a new top-level directory will allow a clear migration path
+for in-repo addons, in the same way that there's a clear migration path from
+`app` to `src`.
+
+Inside `packages`, packages should be grouped by name. Each package can have
+its own `index.js`, `package.json`, and `src` directory.
+
+## Ember Resolver Refactor
+
+The Ember resolver must be refactored significantly to be made aware of the
+new `src` and `packages` directories and associated conventions.
+
+### Module Normalization
+
+As discussed above, Ember CLI will perform a normalization process for all the
+modules in a project and its associated projects. The normalization step will
+involve the construction of a map from each module's normalized form to its
+corresponding ES module path. If any conflicts are detected, the process should
+error and notify the developer.
+
+The Ember resolver will only look up modules in their normalized form, utilizing
+the pre-built normalization map to resolve the actual module path.
+
+### Addon modules
+
+A resolver will only implicitly consider an addon's top-level modules named
+`main` (e.g. a `main` component) to be public and available for resolution. More
+explicit control over an addon's public modules can be declared in the addon's
+`main` module (details TBD). An addon's public modules will all be resolvable at
+the `rootName` of the addon (see above).
+
+Public components and helpers can be invoked in templates using the `rootName`
+as a namespace. For modules named `main`, the bare root name will suffice.
+
+Let's say that the `ember-power-select` addon has a `rootName` of `power-select`
+and a top-level `main` component declared in `src/ui/components/main.js`. An
+app could invoke this component in a template as `{{power-select::main}}` or
+more simply as `{{power-select}}`.
```

Comment:
It seems like `{{power-select}}` could lead to collisions? If so, I'm not in favor of this simplifying syntax.